crates/curp/src/server/raw_curp/log.rs

```diff
@@ -113,19 +120,51 @@ impl<C: Command> LogEntryVecDeque<C> {
     fn push_back(&mut self, entry: Arc<LogEntry<C>>) -> Result<(), bincode::Error> {
         let entry_size = serialized_size(&entry)?;
 
+        if entry_size > self.batch_limit {
+            warn!("entry_size of an entry > batch_limit, which may be too small.",);
+        }
+
         self.entries.push_back(entry);
-        let Some(&pre_entries_size) = self.batch_index.back() else {
-            unreachable!("batch_index cannot be None")
-        };
-        self.batch_index
-            .push_back(pre_entries_size.overflow_add(entry_size));
+        self.entry_size.push_back(entry_size);
+        self.batch_index.push_back(0); // placeholder
+        self.last_batch_size += entry_size;
+
+        while self.last_batch_size > self.batch_limit {
+            if let Some(cur_batch_index) = self.batch_index.get_mut(self.first_entry_at_last_batch)
+            {
+                *cur_batch_index = self.entries.len() - 1 - self.first_entry_at_last_batch;
+                if let Some(cur_entry_size) = self.entry_size.get(self.first_entry_at_last_batch) {
+                    self.last_batch_size -= *cur_entry_size;
+                }
+                self.first_entry_at_last_batch += 1;
+            }
+        }
+
         Ok(())
     }
 
     /// pop a log entry from the front of queue
     fn pop_front(&mut self) -> Option<Arc<LogEntry<C>>> {
         if self.entries.front().is_some() {
-            _ = self.batch_index.pop_front();
+            let front_size = *self
+                .entry_size
+                .front()
+                .unwrap_or_else(|| unreachable!("The entry_size cannot be empty"));
+
+            if self.first_entry_at_last_batch == 0 {
+                self.last_batch_size -= front_size;
+            } else {
+                self.first_entry_at_last_batch -= 1;
+            }
+
+            let _ = self
+                .batch_index
+                .pop_front()
+                .unwrap_or_else(|| unreachable!());
```

Comment:
Please add some messages in this `unreachable!` macro.

Reply:
Done.